### PR TITLE
feat: Service Worker Network Instrumentation and Inspection

### LINKED ISSUE #1090: [Feature] support route aka request interception for service workers
Hi,

I'd like to be able to include markup directly instead of using `page.goto` to a remote address. I currently have a small example here of what I've tried out but it appears that `'serviceWorker' in navigator` is returning false and registering a service worker fails.

I've tried `page.setContent` as well and I had the same results.

Can anyone tell me what I'm missing? The example I am running:

```js
const { chromium } = require('playwright');

const msg = 'hi';

const serviceWorkerPath = '/sw.js';
const serviceWorkerScope = '/';
const serviceWorker = `console.log("${msg}");`;

const html = `<!doctype html>
<html>
  <head>
    <title>One Service Worker</title>
  </head>

  <body>
    <script type="text/javascript">
      console.log('serviceWorker' in navigator);

      try {
        navigator.serviceWorker.register('${serviceWorkerPath}', {
          scope: '${serviceWorkerScope}'
        });
      } catch (e) {
        console.error(e);
      }
    </script>
  </body>
</html>
`.trim();

(async () => {
  const browser = await chromium.launch();
  const page = await browser.newPage();

  await page.route('**/*.html', async request =>
    request.fulfill({
      status: 200,
      contentType: 'text/html',
      body: html,
    }),
  );

  await page.route(serviceWorkerPath, request =>
    request.fulfill({
      status: 200,
      contentType: 'text/javascript',
      body: serviceWorker,
    }),
  );

  await page.goto('http://example.com/index.html', {
    waitUntil: 'load',
  });

  const sw = await page.evaluate(`window.navigator.serviceWorker`);
  
  console.log(sw);

  if (sw) {
    const swTarget = await browser.waitForTarget(target => target.type() === 'service_worker');
    const worker = await browser.serviceWorker(swTarget);
    // use worker target/handle
  }

  await browser.close();
})();
```

### PR DESCRIPTION
**Pull request by rwoll**
Adds Chromium support for Service Worker Networking (interception/routing, Request/Response events, and HAR).

Landed upstream patches:
- https://chromium-review.googlesource.com/c/chromium/src/+/3510917
- https://chromium-review.googlesource.com/c/chromium/src/+/3526571
- https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/3566669
- https://chromium-review.googlesource.com/c/chromium/src/+/3544685
- https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/3610924
- https://chromium-review.googlesource.com/c/chromium/src/+/3689949

TODO: 
- [ ] document update behavior (https://crbug.com/1334900, https://crbug.com/1335463)

Resolves #1090.
Depends on #14714 and #14714.
Supercedes #14321.

### CODE REVIEW

File: tests/library/chromium/chromium.spec.ts [hunk 1]
```diff
@@ -30,6 +30,277 @@ test('should create a worker from a service worker', async ({ page, server }) =>
...
+  ]);
+  expect(innerPage.serviceWorker()).toBe(null);
+  expect((await innerPage.response()).fromServiceWorker()).toBe(true);
+
+  expect(innerSW.serviceWorker()).toBe(worker);
+  expect((await innerSW.response()).fromServiceWorker()).toBe(false);
+});
+
+test('should intercept service worker requests (main and within)', async ({ context, page, server, browserMajorVersion }) => {
+  test.skip(browserMajorVersion < 103, 'Requires fix from https://chromium-review.googlesource.com/c/chromium/src/+/3544685');
```
[the remark this answers is not in the document]
Comment (rwoll) [reply tests/library/chromium/chromium.spec.ts:86]: Yes—looks like it made it into [103.0.5031.0](https://storage.googleapis.com/chromium-find-releases-static/ab1.html#ab1268c40421a390b722774f40f103cf4b2b04cc), and all the stable channels are pointed beyond that: https://omahaproxy.appspot.com/

Thanks!

**Review by dgozman — APPROVED** [review]

File: packages/playwright-core/src/server/chromium/crServiceWorker.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      const r = new network.Route(request, route);
+      if (this._browserContext._requestInterceptor) {
+        this._browserContext._requestInterceptor(r, request);
+        return;
+      }
+      r.continue();
+    }
+  }
+
+  private _isNetworkInspectionEnabled(): boolean {
```
[the remark this answers is not in the document]
Comment (rwoll) [reply packages/playwright-core/src/server/chromium/crServiceWorker.ts:118]: Yes

**Comment by rwoll** [comment]
This is ready to land. However, we're still iterating on har bits for some post-1.23 follow up, so it doesn't make sense to land this quite yet as it will make patching 1.23 quite a nuisance due to the HAR overlap. Once 1.23 settles a bit more, we will land this.

**Comment by rwoll** [comment]
Reviewed CQ1 status, and got verbal LGTM.

**Pull request merged**